From 04be8cb7d619d58c2d89a12b5385a5336d2b32f6 Mon Sep 17 00:00:00 2001
From: Konrad Leszczynski <konrad.leszczynski@intel.com>
Date: Wed, 16 Mar 2016 11:18:23 +0000
Subject: [PATCH 22/22] usb: hub: check for multiple SET_FEATURE requests

SET_FEATURE request for U1/U2 should be issued only
once for each port_dev. Currently, the xHCI tries
to do it multiple times, which could result in
device not being enumerated correctly.

Change-Id: I4fd7324b2e0c26400fae61e0b8ae668fd61c7b43
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-29488
Signed-off-by: Konrad Leszczynski <konrad.leszczynski@intel.com>
Reviewed-on: https://android.intel.com:443/482362
Reviewed-on: https://android.intel.com:443/495742
---
 drivers/usb/core/hub.c | 23 +++++++++++++++++++++--
 drivers/usb/core/hub.h |  3 +++
 2 files changed, 24 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/core/hub.c b/drivers/usb/core/hub.c
index 0ad98c1..b2f039e 100644
--- a/drivers/usb/core/hub.c
+++ b/drivers/usb/core/hub.c
@@ -3924,6 +3924,8 @@ static int usb_disable_link_state(struct usb_hcd *hcd, struct usb_device *udev,
 int usb_disable_lpm(struct usb_device *udev)
 {
 	struct usb_hcd *hcd;
+	struct usb_hub *hub;
+	struct usb_port *port_dev;
 
 	if (!udev || !udev->parent ||
 			udev->speed != USB_SPEED_SUPER ||
@@ -3938,12 +3940,25 @@ int usb_disable_lpm(struct usb_device *udev)
 	if ((udev->u1_params.timeout == 0 && udev->u2_params.timeout == 0))
 		return 0;
 
+	hub = usb_hub_to_struct_hub(udev->parent);
+	if (!hub) {
+		dev_err(&udev->dev, "can't disable lpm, usb_hub is null.\n");
+		return 0;
+	}
+	port_dev = hub->ports[udev->portnum - 1];
+
+	if (port_dev->lpm_disabled) {
+		dev_err(&udev->dev, "lpm already disabled.\n");
+		return 0;
+	}
+
 	/* If LPM is enabled, attempt to disable it. */
 	if (usb_disable_link_state(hcd, udev, USB3_LPM_U1))
 		goto enable_lpm;
 	if (usb_disable_link_state(hcd, udev, USB3_LPM_U2))
 		goto enable_lpm;
 
+	port_dev->lpm_disabled = true;
 	return 0;
 
 enable_lpm:
@@ -4007,11 +4022,15 @@ void usb_enable_lpm(struct usb_device *udev)
 	}
 	port_dev = hub->ports[udev->portnum - 1];
 
-	if (port_dev->u1_allowed)
+	if (port_dev->u1_allowed && !port_dev->u1_enabled) {
 		usb_enable_link_state(hcd, udev, USB3_LPM_U1);
+		port_dev->u1_enabled = true;
+	}
 
-	if (port_dev->u2_allowed)
+	if (port_dev->u2_allowed && !port_dev->u2_enabled) {
 		usb_enable_link_state(hcd, udev, USB3_LPM_U2);
+		port_dev->u2_enabled = true;
+	}
 }
 EXPORT_SYMBOL_GPL(usb_enable_lpm);
 
diff --git a/drivers/usb/core/hub.h b/drivers/usb/core/hub.h
index de57967c..5d7822b 100644
--- a/drivers/usb/core/hub.h
+++ b/drivers/usb/core/hub.h
@@ -98,6 +98,9 @@ struct usb_port {
 	unsigned did_runtime_put:1;
 	unsigned u1_allowed:1;
 	unsigned u2_allowed:1;
+	unsigned u1_enabled:1;
+	unsigned u2_enabled:1;
+	unsigned lpm_disabled:1;
 };
 
 #define to_usb_port(_dev) \
-- 
1.9.1

